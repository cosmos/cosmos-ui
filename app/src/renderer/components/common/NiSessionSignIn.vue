<template lang="pug">
.ni-session: form-struct(:submit='onSubmit').ni-session-container
  .ni-session-header
    a(@click="setState('welcome')"): i.material-icons arrow_back
    .ni-session-title Sign In
    a(@click="help"): i.material-icons help_outline
  .ni-session-main
    form-group(field-id='sign-in-name' field-label='Account Name')
      field#sign-in-name(
        type="text"
        v-model="fields.signInName")
      form-msg(name='Name' type='required' v-if='!$v.fields.signInName.required')

    form-group(:error='$v.fields.signInPassword.$error'
      field-id='sign-in-password' field-label='Password')
      field#sign-in-password(
        type="password"
        v-model="fields.signInPassword")
      form-msg(name='Password' type='required' v-if='!$v.fields.signInPassword.required')
      form-msg(name='Password' type='minLength' min="10" v-if='!$v.fields.signInPassword.minLength')
  .ni-session-footer
    btn(icon="arrow_forward" icon-pos="right" value="Next" size="lg")
</template>

<script>
import {required, minLength} from 'vuelidate/lib/validators'
import Btn from '@nylira/vue-button'
import Field from '@nylira/vue-field'
import FieldGroup from 'common/NiFieldGroup'
import FormGroup from 'common/NiFormGroup'
import FormMsg from '@nylira/vue-form-msg'
import FormStruct from 'common/NiFormStruct'
export default {
  name: 'ni-session-sign-in',
  components: {
    Btn,
    Field,
    FieldGroup,
    FormGroup,
    FormMsg,
    FormStruct
  },
  data: () => ({
    fields: {
      signInName: '',
      signInPassword: ''
    }
  }),
  methods: {
    help () { this.$store.commit('setModalHelp', true) },
    setState (value) { this.$store.commit('setModalSessionState', value) },
    async onSubmit () {
      this.$v.$touch()
      if (this.$v.$error) return
      try {
        await this.$store.dispatch('testLogin', { password: this.fields.signInPassword, account: this.fields.signInName })
        this.$store.dispatch('signIn', { password: this.fields.signInPassword, account: this.fields.signInName })
        this.$store.commit('setModalSession', false)
        this.$store.commit('notify', { title: 'Signed In', body: `You are now signed in to your Cosmos account.` })
      } catch (err) {
        this.$store.commit('notifyError', { title: 'Signing In Failed', body: err.message })
      }
    }
  },
  mounted () {
    this.$el.querySelector('#sign-in-name').focus()
  },
  validations: () => ({
    fields: {
      signInName: { required },
      signInPassword: { required, minLength: minLength(10) }
    }
  })
}
</script>
<style lang="stylus">
  .ni-form-group
    a
      cursor pointer
</style>
