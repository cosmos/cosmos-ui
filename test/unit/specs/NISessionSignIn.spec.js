import Vuex from 'vuex'
import Vuelidate from 'vuelidate'
import { mount, createLocalVue } from 'vue-test-utils'
import htmlBeautify from 'html-beautify'
import NiSessionSignIn from 'common/NiSessionSignIn'

const user = require('renderer/vuex/modules/user').default({})

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuelidate)

describe('NiSessionSignIn', () => {
  let wrapper, store

  beforeEach(() => {
    store = new Vuex.Store({
      modules: {
        user
      }
    })
    wrapper = mount(NiSessionSignIn, {
      localVue,
      store
    })
    store.commit = jest.fn()
    store.dispatch = jest.fn(async () => null)
  })

  it('has the expected html structure', () => {
    expect(htmlBeautify(wrapper.html())).toMatchSnapshot()
  })

  it('should open the help modal on click', () => {
    wrapper.findAll('.ni-session-header a').at(1).trigger('click')
    expect(store.commit).toHaveBeenCalledWith('setModalHelp', true)
  })

  it('should close the modal on successful login', async () => {
    wrapper.setData({ fields: {
      signInPassword: '1234567890',
      signInName: 'name'
    }})
    await wrapper.vm.onSubmit()
    let calls = store.commit.mock.calls.map(args => args[0])
    expect(calls).toContain('setModalSession')
  })

  it('should signal signedin state on successful login', async () => {
    wrapper.setData({ fields: {
      signInPassword: '1234567890',
      signInName: 'name'
    }})
    await wrapper.vm.onSubmit()
    let commitCalls = store.commit.mock.calls.map(args => args[0])
    expect(commitCalls).toContain('notify')
    expect(store.dispatch).toHaveBeenCalledWith('signIn', {
      password: '1234567890',
      account: 'name'
    })
  })

  it('should show error if password not 10 long', () => {
    wrapper.setData({ fields: {
      signInPassword: '123'
    }})
    wrapper.vm.onSubmit()
    expect(store.commit.mock.calls[0]).toBeUndefined()
    expect(wrapper.find('.ni-form-msg-error')).toBeDefined()
  })

  it('should show a notification if signin failed', async () => {
    store.dispatch = jest.fn(() => Promise.reject('Planned rejection'))
    wrapper.setData({ fields: {
      signInPassword: '1234567890',
      signInName: 'name'
    }})
    await wrapper.vm.onSubmit()
    expect(store.commit).toHaveBeenCalled()
    expect(store.commit.mock.calls[0][0]).toBe('notifyError')
  })
})
