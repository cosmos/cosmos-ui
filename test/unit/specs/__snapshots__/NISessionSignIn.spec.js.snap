// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NiSessionSignIn has the expected html structure 1`] = `
"<div class=\\"ni-session\\">
   <form class=\\"ni-session-container ni-form\\">
      <div class=\\"ni-form-container\\">
         <!---->
         <main class=\\"ni-form-main\\">
            <div class=\\"ni-session-header\\">
               <a><i class=\\"material-icons\\">arrow_back</i></a>
               <div class=\\"ni-session-title\\">
                  Sign In
               </div>
      <a><i class=\\"material-icons\\">help_outline</i></a>
   </div>
      <div class=\\"ni-session-main\\">
         <div class=\\"ni-form-group\\">
            <label for=\\"sign-in-name\\">Account Name</label>
            <div class=\\"ni-form-group-field\\">
               <input type=\\"text\\" class=\\"ni-field\\" id=\\"sign-in-name\\">
               <div class=\\"ni-form-msg ni-form-msg-error\\">
                  Name is required
               </div>
</div>
</div>
      <div class=\\"ni-form-group\\">
         <label for=\\"sign-in-password\\">Password</label>
         <div class=\\"ni-form-group-field\\">
            <input type=\\"password\\" class=\\"ni-field\\" id=\\"sign-in-password\\">
            <div class=\\"ni-form-msg ni-form-msg-error\\">
               Password is required
            </div>
               <!---->
   </div>
</div>
</div>
      <div class=\\"ni-session-footer\\">
         <button class=\\"ni-btn\\"><span class=\\"ni-btn-container ni-btn-icon-right ni-btn-size-lg\\"><i aria-hidden=\\"true\\" class=\\"ni-btn-icon material-icons\\">arrow_forward</i><span class=\\"ni-btn-value\\">Next</span></span></button>
      </div>
</main>
      <!---->
</div>
</form>
</div>"
`;
